include/seqan3/alphabet/concept.hpp

```diff
@@ -14,24 +14,668 @@
+ *   3. A member function called `to_rank()`.
+ *      It must be marked `noexcept` (`constexpr` is not required, but recommended) and the return type be
+ *      of the respective rank representation.
+ *
+ * Every (semi-)alphabet type must provide one of the above.
+ *
+ * ### Example
+ *
+ * \include test/snippet/alphabet/to_rank.cpp
+ *
+ * For an example of a full alphabet definition with free function implementations (solution 1. above),
+ * see seqan3::Alphabet.
+ *
+ * ### Customisation point
+ *
+ * This is a customisation point (TODO link to manual). To specify the behaviour for your own alphabet type,
+ * simply provide one of the three functions specified above.
+ */
+inline constexpr auto to_rank = detail::adl::only::to_rank_fn{};
+//!\}
+
+//!\brief The `rank_type` of the semi-alphabet; defined as the return type of seqan3::to_rank.
+//!\ingroup alphabet
+template <typename semi_alphabet_type>
+//!\cond
+    requires requires { { ::seqan3::to_rank(std::declval<semi_alphabet_type>()) }; }
+//!\endcond
+using alphabet_rank_t = decltype(::seqan3::to_rank(std::declval<semi_alphabet_type>()));
+
+} // namespace seqan3
+
+// ============================================================================
+// assign_rank_to()
+// ============================================================================
+
+namespace seqan3::detail::adl::only
+{
+
+//!\brief Functor definition for seqan3::assign_rank_to.
+//!\ingroup alphabet
+struct assign_rank_to_fn
+{
+private:
+    SEQAN3_CPO_IMPL(2, (assign_rank_to(args..., v)                           ))    // ADL
+    SEQAN3_CPO_IMPL(1, (seqan3::adaptation::assign_rank_to(args..., v)       ))    // customisation namespace
+    SEQAN3_CPO_IMPL(0, (v.assign_rank(args...)                               ))    // member
+
+public:
+    //!\brief Operator definition for lvalues.
+    template <typename alph_t>
+    constexpr auto operator()(seqan3::alphabet_rank_t<alph_t> const r, alph_t & a) const noexcept
+    //!\cond
+        -> decltype(impl(priority_tag<2>{}, a, r))
+    //!\endcond
+    {
+        static_assert(noexcept(impl(priority_tag<2>{}, a, r)),
+            "Only overloads that are marked noexcept are picked up by seqan3::assign_rank_to().");
+        static_assert(std::Same<alph_t &, decltype(impl(priority_tag<2>{}, a, r))>,
+            "The return type of your assign_rank_to() implementation must be 'alph_t &'.");
+
+        return impl(priority_tag<2>{}, a, r);
+    }
+
+    //!\brief Operator definition for rvalues.
+    template <typename alph_t>
+    //!\cond
+        requires (!std::is_lvalue_reference_v<alph_t>)
+    //!\endcond
+    constexpr auto operator()(seqan3::alphabet_rank_t<alph_t> const r, alph_t && a) const noexcept
+    //!\cond
+        -> std::remove_reference_t<decltype(impl(priority_tag<2>{}, a, r))>
+    //!\endcond
+    {
+        return (*this)(r, a); // call above function but return by value
```

Comment:
I would prefer the syntax: this->operator()(r, a); as it is easier to read. But I have no strong feelings.

Reply:
Hm, I wanted to do `operator()(r, a)` but that does self-invocation.  The above solution was the shortest and it is even commented?

Reply:
It still looks weird, but as I sad, no strong feelings.